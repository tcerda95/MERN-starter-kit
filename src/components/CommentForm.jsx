import React, { Component } from 'react';
import { translate } from 'react-i18next';
import PropTypes from 'prop-types';
import 'bulma/css/bulma.css';

export class CommentForm extends Component {
  static propTypes = {
    onCommentSubmit: PropTypes.func.isRequired,
    onLogout: PropTypes.func.isRequired,
    imageURL: PropTypes.string
  }

  state = { text: '' }

  handleTextChange = e => {
    this.setState({ text: e.target.value });
  }

  handleKeyPress = e => {
    if (e.key === 'Enter')
      this.handleSubmit(e);
  }

  handleSubmit = e => {
    e.preventDefault();
    const text = this.state.text.trim();
    
    if (!text)
      return;

    this.props.onCommentSubmit({ text });
    this.setState({ text: '' });
  }

  handleLogout = e => {
    e.preventDefault();
    this.props.onLogout();
  }

  render() {
    const { t } = this.props;
    
    return (
      <article className="media">
        <figure className="media-left">
          <p className="image is-64x64">
            <img alt="Avatar" src={this.props.imageURL} />
          </p>
        </figure>
        <div className="media-content">
          <div className="field">
            <p className="control">
              <textarea
                className="textarea"
                placeholder={t('addComment')}
                value={this.state.text}
                onChange={this.handleTextChange}
                onKeyPress={this.handleKeyPress}
              />
            </p>
          </div>

          <nav className="level">
            <div className="level-left">
              <div className="level-item">
                <a
                  className="button is-info"
                  type="submit"
                  value={t('post')}
                  onClick={this.handleSubmit}
                >
                  {t('post')}
                </a>
              </div>
            </div>

            <div className="level-right">
              <div className="level-item">
                <a
                  className="button is-info"
                  type="submit"
                  value={t('logout')}
                  onClick={this.handleLogout}
                >
                  {t('logout')}
                </a>
              </div>
            </div>
          </nav>
        </div>
      </article>
    );
  }
}

export default translate()(CommentForm);
