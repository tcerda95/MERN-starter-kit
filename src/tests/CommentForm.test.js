import React from 'react';
import sinon from 'sinon';
import CommentForm from '../components/CommentForm';
import propTypesUtils from './utils/propTypes-utils.js';
import { shallow } from 'enzyme';

describe('React CommentForm testing', () => {
  const { assert } = sinon;
  let onCommentSubmitSpy;
  let onLogoutSpy;
  let preventDefaultSpy;
  let wrapper;

  beforeEach(() => {
    onCommentSubmitSpy = sinon.stub();
    onLogoutSpy = sinon.spy();
    preventDefaultSpy = sinon.spy();
    wrapper = shallow(<CommentForm t={t} onCommentSubmit={onCommentSubmitSpy} onLogout={onLogoutSpy} />);
  });

  it('should raise an error if any required props is missing', () => {
    propTypesUtils.spyConsoleError();

    shallow(<CommentForm t={t} />);
    propTypesUtils.expectMissingProp('onCommentSubmit', 'CommentForm');
    propTypesUtils.expectMissingProp('onLogout', 'CommentForm');

    propTypesUtils.restoreConsoleError();
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have value "a" on keypress "a"', () => {
    let input = wrapper.find('textarea');
    input.simulate('change', { target: { value: 'testing...' } });

    input = wrapper.find('textarea');
    expect(input.get(0).props).toHaveProperty('value', 'testing...');
  });

  it('should start with no input', () => {
    const input = wrapper.find('textarea');
    expect(input.get(0).props).toHaveProperty('value', '');    
  });

  it('should submit on keypress Enter if there is content', () => {
    const input = wrapper.find('textarea');
    input.simulate('change', { target: { value: 'testing...' } });
    input.simulate('keypress', { key: 'Enter', preventDefault: preventDefaultSpy });

    assert.calledOnce(onCommentSubmitSpy);
    assert.calledWith(onCommentSubmitSpy, { text: 'testing...' });
    assert.calledOnce(preventDefaultSpy);
  });

  it('should not submit on keypress Enter if there is no content', () => {
    const input = wrapper.find('textarea');
    input.simulate('keypress', { key: 'Enter', preventDefault: preventDefaultSpy });

    assert.notCalled(onCommentSubmitSpy);
    assert.calledOnce(preventDefaultSpy);
  });
});
