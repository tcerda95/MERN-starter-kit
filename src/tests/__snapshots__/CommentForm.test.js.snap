// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React CommentForm testing should match snapshot 1`] = `
<article
  className="media"
>
  <figure
    className="media-left"
  >
    <p
      className="image is-64x64"
    >
      <img
        alt="Avatar"
      />
    </p>
  </figure>
  <div
    className="media-content"
  >
    <div
      className="field"
    >
      <p
        className="control"
      >
        <textarea
          className="textarea"
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="addComment"
          value=""
        />
      </p>
    </div>
    <nav
      className="level"
    >
      <div
        className="level-left"
      >
        <div
          className="level-item"
        >
          <a
            className="button is-info"
            onClick={[Function]}
            type="submit"
            value="post"
          >
            post
          </a>
        </div>
      </div>
      <div
        className="level-right"
      >
        <div
          className="level-item"
        >
          <a
            className="button is-info"
            onClick={[Function]}
            type="submit"
            value="logout"
          >
            logout
          </a>
        </div>
      </div>
    </nav>
  </div>
</article>
`;
